typescript-rewrite/samlang-core/src/interpreter/expression-interpreter.ts

```diff
@@ -0,0 +1,276 @@
+            const v2 = this.eval(expression.e2) as bigint;
+            return v1 > v2;
+          }
+          case '>=': {
+            const v1 = this.eval(expression.e1) as bigint;
+            const v2 = this.eval(expression.e2) as bigint;
+            return v1 >= v2;
+          }
+          case '==': {
+            const v1 = this.eval(expression.e1);
+            const v2 = this.eval(expression.e2);
+            if (
+              (v1 as FunctionValue).type === 'functionValue' ||
+              (v2 as FunctionValue).type === 'functionValue'
+            ) {
+              throw new PanicException('Cannot compare functions!');
+            }
+            return v1 === v2;
+          }
+          case '!=': {
+            const v1 = this.eval(expression.e1);
+            const v2 = this.eval(expression.e2);
+            if (
+              (v1 as FunctionValue).type === 'functionValue' ||
+              (v2 as FunctionValue).type === 'functionValue'
+            ) {
+              throw new PanicException('Cannot compare functions!');
+            }
+            return v1 !== v2;
+          }
+          case '&&': {
+            const v1 = this.eval(expression.e1) as boolean;
+            return v1 && this.eval(expression.e2, context);
+          }
+          case '||': {
+            const v1 = this.eval(expression.e1) as boolean;
+            return v1 || this.eval(expression.e2, context);
+          }
+          case '::': {
+            const v1 = this.eval(expression.e1) as string;
+            const v2 = this.eval(expression.e2) as string;
+            return v1 + v2;
+          }
+        }
+      }
+      // eslint-disable-next-line no-fallthrough
+      case 'IfElseExpression': {
+        return this.eval(
+          (this.eval(expression.boolExpression) as boolean) ? expression.e1 : expression.e2,
+          context
+        );
+      }
+      case 'MatchExpression': {
+        const matchedValue = this.eval(expression.matchedExpression, context) as VariantValue;
+        const matchedPattern =
+          expression.matchingList.find((el) => el.tag === matchedValue.tag) ??
+          this.blameTypeChecker();
+        let ctx = context;
+        if (matchedPattern.dataVariable) {
+          ctx = {
+            classes: ctx.classes,
+            localValues: { ...ctx.localValues, [matchedPattern.dataVariable]: matchedValue.data },
+          };
+        }
+        return this.eval(matchedPattern.expression, ctx);
+      }
+      case 'LambdaExpression':
+        return {
+          type: 'functionValue',
+          arguments: expression.parameters.map((param) => param[0]),
+          body: expression.body,
+          context,
+        };
+      case 'StatementBlockExpression': {
+        const { block } = expression;
+        const localValues = { ...context.localValues };
+        block.statements.forEach((statement) => {
+          const assignedValue = this.eval(statement.assignedExpression, context);
```

Comment:
You probably want something like

```ts
// safe to leave classes along since they are immutable since initialization.
const contextForStatementBlock = {...context, localValues: {...context.localValues} }
```

and pass contextForStatementBlock to eval in line 244 and later similar lines

Reply:
Added in latest commit... but also realized I never actually passed the context to evaluate the final expression 😅 the nest statement block test should be good to test for this case 👍